linkml/generators/pythongen.py

```diff
@@ -769,6 +774,38 @@ def is_key_value_class(self, range_name: DefinitionName) -> bool:
                 return len(rng.slots) - len(pkeys) == 1
         return False
 
+    def gen_constructor(self, cls: ClassDefinition) -> Optional[str]:
+        rlines: List[str] = []
+        designators = [x for x  in self.domain_slots(cls) if x.designates_type]
+        if len(designators) > 0:
+            descendants = self.schemaview.class_descendants(cls.name)
+            if len(descendants) > 1:
+                slot = designators[0]
+                aliased_slot_name = self.slot_name(
+                    slot.name
+                )
+                rlines.append("def __new__(cls, *args, **kwargs):")
+                rlines.append(f"\tmappings_{aliased_slot_name} = {{")
+                for descendant in descendants:
+                    d = self.schema.classes[descendant]
+                    descendant_class_tv = get_type_designator_value(self.schemaview, slot, d)
+                        # the URI defined in the schemaview is empty so i need to take it out of the schemaloader.
+                        # this seems suboptimal since this will eventually move to schemaview
+                    rlines.append(f"\t\t\"{descendant_class_tv}\": {self.class_or_type_name(d.name)},")
+                rlines.append(f"\t}}")
+
+                rlines.append(f"""
+        type_designator = "{aliased_slot_name}"
+        if not type_designator in kwargs:
+            return super().__new__(cls,*args,**kwargs)
```

Comment:
We're a little nervous about whether we have enough tests that take advantage of designates_type to catch unintended side effects. One thought I had was that even though this is likely perfectly safe to just pass up to the super constructor without arguments, it might be a step safer to just return an empty string in the case that none of the subclasses has a designates_type slot. 

(I think maybe on/before line 782 you might be able to use a list comprehension to count the number of descendants with designates_type slots, rather than the total number of descendants)

Reply:
Hello! I'm not sure if i understood everything correctly..

For the list comprehension: don't all child classes of a class having a type designator have a type designator too ? Would there be a chance that the list comprehension would give less classes than the full class descendants list ?

Btw, for the pydantic generator i did the same thing (take full list of class descendants to build the Union[])..

Thanks!
Frank

Reply:
In any case, i will add an extra test that verifies the correct working if we create an instance of a class having a type designator, without actually passing the value of the type designator (this used to work before, so it should still work).

Reply:
the extra test uncovered some issues:

* I only generate a __new__ when the type designator is a class slot and not an induced slot. This is wrong and breaks if i just instantiate an object of a subclass: then the type designator is not initialized correctly.

* When i have a class hierarchy A --> B --> C and i instantiate a B that is actually a C (because of its type designator value), it remains at B because the type designator was defined at A and so only A has a __new__

going to fix these